clang/lib/CodeGen/BackendUtil.cpp

```diff
@@ -915,12 +915,10 @@ void EmitAssemblyHelper::RunOptimizationPipeline(
 
   ModulePassManager MPM;
 
-  // FIXME: Change this when -fno-sycl-early-optimizations is not tied to
-  // -disable-llvm-passes.
-  if (CodeGenOpts.DisableLLVMPasses && LangOpts.SYCLIsDevice)
-    MPM.addPass(SYCLPropagateAspectsUsagePass());
-
   if (!CodeGenOpts.DisableLLVMPasses) {
+    if (LangOpts.SYCLIsDevice)
+      MPM.addPass(SYCLPropagateAspectsUsagePass());
```

Comment:
The `SYCLPropagateAspectsUsagePass` is added in early passes a couple lines down. Previously we needed to add it if `CodeGenOpts.DisableLLVMPasses` was true to work around `-fno-sycl-early-optimizations` being analogous to `-disable-llvm-passes`, but now it should be enough to just have the one further down. As such, I believe this can be removed.

Reply:
Just tested and removing just those two lines causes the following test failures:

```
Failed Tests (2):
  Clang :: CodeGenSYCL/sub-group-size.cpp
  SYCL :: check_device_code/device_has.cpp
```
- `sub-group-size.cpp` fails on [line 18](https://github.com/intel/llvm/blob/a246cdb524ca7c288f0a7ecd76cf62640414c8b9/clang/test/CodeGenSYCL/sub-group-size.cpp#L18) as the NONE-DAG check doesn't match - there is no `sycl_fixed_targets` attribute on the emitted `external_default_behavior` function bytecode.
- `func1` in the `device_has.cpp` test doesn't get decorated with a `!sycl_used_aspects` attribute, causing the test to fail at [line 13](https://github.com/intel/llvm/pull/7701/files#diff-2de6fffb6430b0e1e18dea3e91afda09ac510d2b90771f10d0c4b25a38aedb6e) (the `CHECK_SAME` test).

Reply:
It looks like those callbacks are never called if we don't use a predefined method for building a pipeline.

I suggest that we then always build some pipeline, but when early optimizations are disabled, we build `O0DefaultPipeline` - that will trigger all callback functions we have and insert SYCL-specific passed, but won't do any optimizations. Please note that `buildO0DefaultPipeline` contains an `assert` that optimization level is `O0`, so we will have to explicitly pass `OptimizationLevel::O0` there if early optimizations are disabled.